memgpt/agent.py

```diff
@@ -912,146 +847,111 @@ def heartbeat_is_paused(self):
         )[0]
+        print("NEW SYSTEM MESSAGE", new_system_message["content"])
 
         diff = united_diff(curr_system_message["content"], new_system_message["content"])
-        printd(f"Rebuilding system with new memory...\nDiff:\n{diff}")
+        if len(diff) > 0:  # there was a diff
+            printd(f"Rebuilding system with new memory...\nDiff:\n{diff}")
 
-        # Swap the system message out
-        self._swap_system_message(
-            Message.dict_to_message(
-                agent_id=self.agent_state.id, user_id=self.agent_state.user_id, model=self.model, openai_message_dict=new_system_message
+            # Swap the system message out (only if there is a diff)
+            print("swap system message")
+            self._swap_system_message(
+                Message.dict_to_message(
+                    agent_id=self.agent_state.id, user_id=self.agent_state.user_id, model=self.model, openai_message_dict=new_system_message
+                )
+            )
+            assert self.messages[0]["content"] == new_system_message["content"], (
+                self.messages[0]["content"],
+                new_system_message["content"],
             )
-        )
-
-    # def to_agent_state(self) -> AgentState:
-    #    # The state may have change since the last time we wrote it
-    #    updated_state = {
-    #        "persona": self.memory.persona,
-    #        "human": self.memory.human,
-    #        "system": self.system,
-    #        "functions": self.functions,
-    #        "messages": [str(msg.id) for msg in self._messages],
-    #    }
-
-    #    agent_state = AgentState(
-    #        name=self.agent_state.name,
-    #        user_id=self.agent_state.user_id,
-    #        persona=self.agent_state.persona,
-    #        human=self.agent_state.human,
-    #        llm_config=self.agent_state.llm_config,
-    #        embedding_config=self.agent_state.embedding_config,
-    #        preset=self.agent_state.preset,
-    #        id=self.agent_state.id,
-    #        created_at=self.agent_state.created_at,
-    #        state=updated_state,
-    #    )
-
-    #    return agent_state
 
     def add_function(self, function_name: str) -> str:
-        if function_name in self.functions_python.keys():
-            msg = f"Function {function_name} already loaded"
-            printd(msg)
-            return msg
-
-        available_functions = load_all_function_sets()
-        if function_name not in available_functions.keys():
-            raise ValueError(f"Function {function_name} not found in function library")
-
-        self.functions.append(available_functions[function_name]["json_schema"])
-        self.functions_python[function_name] = available_functions[function_name]["python_function"]
-
-        msg = f"Added function {function_name}"
-        # self.save()
-        self.update_state()
-        printd(msg)
-        return msg
+        # TODO: refactor

```

Comment:
probably should `raise NotImplementedError` instead?

Reply:
or raise a `Warning`?

Reply:
```
import warnings
warnings.warn("...")
```

Reply:
I did a `NotImplementedError`